File: Leetcode/java/_3_Longest_Substring_Without_Repeating_Characters.java
// link - https://leetcode.com/problems/longest-substring-without-repeating-characters/

import java.util.HashSet;

public class _3_Longest_Substring_Without_Repeating_Characters {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        int max = 0, curr = 0;
        int i = 0, j = 0;

        while (j < s.length()) {
            if (!set.contains(s.charAt(j))) {
                curr++;
                max = Math.max(max, curr);
                set.add(s.charAt(j));
                j++;
            } else {
                while (set.contains(s.charAt(j))) {
                    set.remove(s.charAt(i));
                    i++;
                    curr--;
                }
            }
        }

        return max;
    }
}

File: Leetcode/python/_3_Longest_Substring_Without_Repeating_Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if s == "": return 0
        seen = set(s[0])
        max_len = 1

        i, j = 0, 1
        while j < len(s) and i < j:
            while s[j] in seen:
                seen.remove(s[i])
                i += 1
            seen.add(s[j])
            max_len = max(max_len, len(seen))
            j += 1

        return max_len
